src/Azure/Azure.Quantum.Client/Authentication/TokenFileCredential.cs

```diff
@@ -0,0 +1,139 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#nullable enable
+
+namespace Microsoft.Azure.Quantum.Authentication
+{
+    using System;
+    using System.Text.Json;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    using global::Azure.Core;
+    using global::Azure.Identity;
+    using Microsoft.Azure.Quantum.Utility;
+
+    /// <summary>
+    /// Implements a custom TokenCredential to use a local file as the source for an AzureQuantum token.
+    ///
+    /// It will only use the local file if the AZUREQUANTUM_TOKEN_FILE environment variable is set, and references
+    /// an existing json file that contains the access_token and expires_on timestamp in milliseconds.
+    ///
+    /// If the environment variable is not set, the file does not exist, or the token is invalid in any way(expired, for example),
+    /// then the credential will throw CredentialUnavailableError, so that DefaultQuantumCredential can fallback to other methods.
+    /// </summary>
+    public class TokenFileCredential : TokenCredential
+    {
+        /// <summary>
+        /// Environment variable name for the token file path.
+        /// </summary>
+        private const string TokenFileEnvironmentVariable = "AZUREQUANTUM_TOKEN_FILE";
+
+        /// <summary>
+        /// File system dependency injected so that unit testing is possible.
+        /// </summary>
+        private readonly IFileSystem _fileSystem;
+
+        /// <summary>
+        /// The path to the token file.
+        /// </summary>
+        private readonly string? _tokenFilePath;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TokenFileCredential"/> class.
+        /// </summary>
+        public TokenFileCredential()
+            : this(new FileSystem())
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TokenFileCredential"/> class.
+        /// </summary>
+        /// <param name="fileSystem">The file system.</param>
+        public TokenFileCredential(IFileSystem fileSystem)
+        {
+            _fileSystem = fileSystem;
+            _tokenFilePath = Environment.GetEnvironmentVariable(TokenFileEnvironmentVariable);
+        }
```

Comment:
[minor] You may be able to eliminate the parameterless constructor above by making `fileSystem` nullable:

```suggestion
        public TokenFileCredential(IFileSystem? fileSystem = null)
        {
            _fileSystem = fileSystem ?? new FileSystem();
            _tokenFilePath = Environment.GetEnvironmentVariable(TokenFileEnvironmentVariable);
        }
```

Reply:
Fixed!